File: Python/0001-0050/24. Swap Nodes in Pairs.py
# https://leetcode.com/problems/swap-nodes-in-pairs/description/

class Solution:
    def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head == None or head.next == None:
            return head

        previous = None
        current = head
        next = current.next

        while next != None:
            # Swap Nodes
            current.next = next.next
            next.next = current
            if previous == None:
                head = next
            else:
                previous.next = next

            # Move Nodes
            previous = current
            current = previous.next
            if current == None:
                break
            next = current.next


        return head

File: Java/0001-0050/24. Swap Nodes in Pairs.java
// https://leetcode.com/problems/swap-nodes-in-pairs/description/

class Solution {
    public ListNode swapPairs(ListNode head) {
        if(head == null || head.next==null){
            return head;
        }
        
        ListNode previous = null;
        ListNode current = head;
        ListNode next = head.next;

        while(next!=null){
            current.next = next.next;
            next.next = current;
            if(previous==null){
                head = next;
            }else{
                previous.next = next;
            }

            //move nodes
            previous = current;
            current = previous.next;
            if(current==null){
                break;
            }
            next = current.next;

        }
        return head;
    }
}
